ext/snowcrash/src/UriTemplateParser.cc

```diff
@@ -9,207 +9,251 @@
-        return queryStringExpression;
+    void operator()(const state::invalid& s) const {
+
+        tao::pegtl::memory_input<> in(s.content, "");
+        error_locator::state state;
+        tao::pegtl::parse<error_locator::grammar<varspec>, error_locator::report_action>(in, state);
+
+        std::stringstream ss;
+        ss << "URI template variable '" << s.content
+           << "' contains invalid " << state.message
+           <<". Allowed characters for expressions are A-Z a-z 0-9 _ and percent encoded characters";
+
+#if 0
+        /**
+         *  this is prepared to set exact location of invalid character
+         *  but in current version we receive souce map for full makdown node instead of position
+         *  URI template so we are not able to set exact location of invalid character
+         */
+        assert(sourceBlock.size() == 1);
+        mdp::CharactersRangeSet err;
+        err.emplace_back(mdp::CharactersRange{sourceBlock.front().location + s.position + state.position, s.content.length()});
+
+        result.report.warnings.push_back(Warning(ss.str(), URIWarning, err));
+#endif
+        result.report.warnings.push_back(Warning(ss.str(), URIWarning, sourceBlock));
     }
 
-    FragmentExpression fragmentExpression(expression);
-    if (fragmentExpression.IsExpressionType()) {
-        return fragmentExpression;
-    }
+};
 
-    ReservedExpansionExpression reservedExpansionExpression(expression);
-    if (reservedExpansionExpression.IsExpressionType()) {
-        return reservedExpansionExpression;
-    }
+struct TemplateValidityChecker {
+    ParsedURITemplate& result;
+    const mdp::CharactersRangeSet& sourceBlock;
 
-    LabelExpansionExpression labelExpansionExpression(expression);
-    if (labelExpansionExpression.IsExpressionType()) {
-        return labelExpansionExpression;
+    void operator()(const state::literals& s) const {
+        // do nothing literal is parsed correctly, so there is nothing to report
     }
 
-    PathSegmentExpansionExpression pathSegmentExpansionExpression(expression);
-    if (pathSegmentExpansionExpression.IsExpressionType()) {
-        return pathSegmentExpansionExpression;
+    void operator()(const state::invalid& s) const {
+
+        tao::pegtl::memory_input<> in(s.content, "");
+        error_locator::state state;
+        tao::pegtl::parse<error_locator::grammar<literals_>, error_locator::report_action>(in, state);
+
+        std::stringstream ss;
+        ss << "URI template contains invalid " << state.message
+           << ". You probably will need to repace invalid chars by percent encoded character.";
+
+#if 0
+        /**
+         *  this is prepared to set exact location of invalid character
+         *  but in current version we receive souce map for full makdown node instead of position
+         *  URI template so we are not able to set exact location of invalid character
+         */
+        assert(sourceBlock.size() == 1);
```

Comment:
Are you sure this should be a debug assertion? Note that if `sourceBlock` is empty, line 168 comes with UB.

Reply:
It is proposal how to correct implement location of invalid character.
It is also reason why it is closed in `#if 0/#endif` block - see comment above.

Reply:
> It is proposal how to correct implement location of invalid character

Given the current behaviour is the same as the one in this PR. I'd descope this change from this changeset so we can move it forward. Improving the source maps can be done as part of a separate task which is already tracked in JIRA.